src/Shared/PackageManagers/BaseProjectManager.cs

```diff
@@ -364,14 +384,21 @@ public virtual async Task<bool> PackageVersionExistsAsync(PackageURL purl, bool
         }
         
         /// <summary>
-        /// Check if the package version was pulled from the repository.
+        /// Check if the package version exists/existed in the repository.
         /// </summary>
         /// <param name="purl">The PackageURL to check.</param>
-        /// <param name="useCache">If the cache should be checked for the existence of this package.</param>
-        /// <returns>True if the package was pulled from the repository. False otherwise.</returns>
-        public virtual Task<bool> PackageVersionPulled(PackageURL purl, bool useCache = true)
+        /// <param name="useCache">If the cache should be checked for the existence of this package version.</param>
+        /// <returns>A <see cref="IPackageExistence"/> detailing the existence of the package version.</returns>
+        public virtual async Task<IPackageExistence> DetailedPackageVersionExistsAsync(PackageURL purl, bool useCache = true)
         {
-            throw new NotImplementedException("BaseProjectManager does not implement PackageVersionPulled.");
+            bool exists = await PackageVersionExistsAsync(purl, useCache);
+
+            if (exists)
+            {
+                return new PackageVersionExists();
+            }
+
+            return new PackageVersionNotFound();
         }
```

Comment:
I was thinking a bit more about the default behavior you have here - in the event that a package version was unpublished, `PackageVersionExistsAsync` should return `false`, correct?

I ask because I want to double check the extensibility of this API contract if/when we add support for `PackageVersionRemoved` going forward. One question I'm thinking about is whether `PackageVersionRemoved` should be a subtype of `PackageVersionNotFound`, or whether they should be independent (as currently written in this PR). 

If `PackageVersionExistsAsync` does in fact return `false` for the unpublished/removed cases, then the default implementation of this method will return `PackageVersionNotFound` for now. When we come back and start adding support for `PackageVersionRemoved` to additional types of packages, it seems like `PackageVersionRemoved` would need to be a subtype of `PackageVersionNotFound`, to avoid breaking any code that relies on the default behavior here of returning `PackageVersionNotFound` in that case (i.e. if a caller has an expression like `result is PackageVersionNotFound`, and that expression evaluated to true for a given unpublished Cargo package version, it should continue to evaluate to true after we add support for `PackageVersionRemoved` for Cargo).

Reply:
Same discussion applies to `DetailedPackageExistsAsync` as well.

Reply:
This makes sense to me. Removed implies it wasn't found so I can make it a subset.

Reply:
Give it a look when you can, but I think I addressed what you were getting at, and added unit tests for it as well!